# Project: Database

**Instructions for Students:**

Please carefully follow these steps to complete and submit your project:

1. **Completing the Project**: You are required to work on and complete all tasks in the provided project. Be disciplined and ensure that you thoroughly engage with each task.
   
2. **Creating a Google Drive Folder**: Each of you must create a new folder on your Google Drive if you haven't already. This will be the repository for all your completed assignment and project files, aiding you in keeping your work organized and accessible.
   
3. **Uploading Completed Project**: Upon completion of your project, make sure to upload all necessary files, involving codes, reports, and related documents into the created Google Drive folder. Save this link in the 'Student Identity' section and also provide it as the last parameter in the `submit` function that has been provided.
   
4. **Sharing Folder Link**: You're required to share the link to your project Google Drive folder. This is crucial for the submission and evaluation of your project.
   
5. **Setting Permission toPublic**: Please make sure your Google Drive folder is set to public. This allows your instructor to access your solutions and assess your work correctly.

Adhering to these procedures will facilitate a smooth project evaluation process for you and the reviewers.

## Student Identity

In [1]:
# @title #### Student Identity
student_id = "REAGRZQD" # @param {type:"string"}
name = "Theofilus Arifin" # @param {type:"string"}
drive_link = "https://drive.google.com/drive/folders/13jcavEC5d6Yq6E46eqrUyg4VZB8NjkOu?usp=sharing"  # @param {type:"string"}

assignment_id = "00_database_project"

# Import grader package
!pip install rggrader
from rggrader import submit, submit_image

## Project Description

In this project, you'll work with the Kaggle E-commerce dataset. Your tasks are to import this dataset into a SQLite database, explore it using SQL to identify key attributes and patterns, and perform basic data cleaning procedures. This project is designed to improve your skills in database management, SQL usage, and data cleaning.

Rmember to make a copy of this notebook in your Google Drive and work in your own copy.

## I. Loading and Querying with SQL

In this task, you will use Kaggle E-commerce dataset. You need to download it from the link here: [Kaggle E-commerce data](https://www.kaggle.com/datasets/carrie1/ecommerce-data?select=data.csv).

After you download the dataset, you will then import the dataset into SQL format, using SQLite.

### I.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [2]:
import sqlite3
import pandas as pd

### I.2. Data Loading and SQL CREATE and INSERT to database table

Now you can create the database and tables. Follow the same column naming as the dataset.

Next insert all the data from the dataset into the table in the database.

In [3]:
!git clone https://github.com/Theofilusarifin/E-Commerce-Data-Analysis-with-Elastic-Search

Cloning into 'E-Commerce-Data-Analysis-with-Elastic-Search'...
remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 7 (delta 0), reused 3 (delta 0), pack-reused 0
Receiving objects: 100% (7/7), 7.23 MiB | 412.00 KiB/s, done.


In [4]:
df = pd.read_csv('/content/E-Commerce-Data-Analysis-with-Elastic-Search/dataset.csv', encoding='latin1')
df.shape

(541909, 8)

In [5]:
df.head(3)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   
2    536365    84406B      CREAM CUPID HEARTS COAT HANGER         8   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom

In [6]:
conn = sqlite3.connect('sqlite.db')

cur = conn.cursor()

#### Create Table

In [7]:
cur.execute('''DROP TABLE IF EXISTS ecommerce''')

In [8]:
cur.execute('''
  CREATE TABLE IF NOT EXISTS ecommerce (
      InvoiceNo VARCHAR,
      StockCode VARCHAR,
      Description TEXT,
      Quantity INTEGER,
      InvoiceDate DATETIME,
      UnitPrice FLOAT,
      CustomerID FLOAT,
      Country VARCHAR
  )
''')

# Commit the transaction
conn.commit()

In [9]:
pd.read_sql_query('SELECT * FROM ecommerce', conn)

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

#### Insert Data

In [10]:
for index, row in df.iterrows():
    cur.execute('''
        INSERT INTO ecommerce (InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
    ''', (row['InvoiceNo'], row['StockCode'], row['Description'], row['Quantity'], row['InvoiceDate'], row['UnitPrice'], row['CustomerID'], row['Country'])
    )

conn.commit()

In [11]:
pd.read_sql_query('SELECT * FROM ecommerce', conn).shape

(541909, 8)

In [12]:
conn.close()

### I.3. Verifying the database

Before we go to the next section, let's check our database first. You need to change the db_name and table_name with your database name and table name, afterwards you can just run the code block. The result should be:

- Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
- Number of rows: 541909

In [13]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('sqlite.db') #change db_name to your database name.
cur = conn.cursor()

# Get column names
cur.execute("PRAGMA table_info(ecommerce)") #change table_name to your table name
sql_columns = [column[1] for column in cur.fetchall()] #the variable sql_columns is used for submission, do not change the variable name
print(f"Columns: {sql_columns}")

# Get total number of rows
cur.execute("SELECT COUNT(*) FROM ecommerce") #change table_name to your table name
sql_num_rows = cur.fetchone()[0] #the variable sql_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {sql_num_rows}")

# Close the connection
conn.close()

Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
Number of rows: 541909


### I.3.4. Running SQL Queries

Once we have the database in SQL format, let's do some data exploration and analysis.

You are part of the engineering team at a global e-commerce company. Your company has a vast catalogue of products and serves customers around the globe. As the company continues to grow, the management decides it's time to expand the business even more.

As an essential part of this expansion, your team is tasked with diving into your company's rich e-commerce dataset to extract crucial insights that will guide the expansion strategy.

1. **Total number of unique products and unique customers:**

   Your first task is to gauge the breadth of your operations. By determining the total number of unique products you sell and the total number of unique customers you serve, you can assess the scale and diversity of your business.

2. **Total revenue for each product:**

   Next, you're going to identify the star performers in your product portfolio. Calculating the total revenue for each product will reveal which items are the biggest revenue drivers. These insights can help guide decisions about product focus and marketing efforts. Limit the result to the top 5 product with the biggest revenue.

3. **Most profitable countries (by total sales):**

   To pinpoint the most fruitful geographic areas for your business, you'll identify which countries generate the most profit. This data will help you understand where your operations are already strong and where there might be potential for regional expansion. Limit the result to the top 5 countries with the biggest profit.

4. **Top-selling products (by total sales) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by sales for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

4. **Top-selling products (by quantity) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by quantity for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

By answering these queries, your team will provide valuable data-driven insights that will directly contribute to the strategic decisions about where and how your business expands. As part of the engineering team, you're not just maintaining the technical infrastructure - you're shaping the future of the company.

> Hint: The Product name is stored in column Descrioption

In [14]:
conn = sqlite3.connect('sqlite.db')
cur = conn.cursor()

In [15]:
# Write your SQL Query here
sql_num_products = 0
sql_num_customers = 0
sql_top_product_revenues = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_countries = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_selling_products_by_country_by_sales = {} #dictionary where key is country name, value is a list of top-selling products by sales, see example below
sql_top_selling_products_by_country_by_quantity = {} #dictionary where key is country name, value is a list of top-selling products by quantity, see example below

#### 1. Find the total number of unique products and unique customers

In [16]:
cur.execute("SELECT COUNT(DISTINCT(StockCode)) FROM ecommerce")
sql_num_products = cur.fetchone()[0]
print(f"Number of Unique Products: {sql_num_products}")

Number of Unique Products: 4070


In [17]:
cur.execute("SELECT COUNT(DISTINCT(CustomerID)) FROM ecommerce")
sql_num_customers = cur.fetchone()[0]
print(f"Number of Unique Customer: {sql_num_customers}")

Number of Unique Customer: 4372


#### 2. Calculate the total revenue for each product, limited to top 5

In [18]:
product_revenue = pd.read_sql_query('''
  SELECT
    Description,
    SUM(Quantity * UnitPrice) as Revenue
  FROM
    ecommerce
  GROUP BY
    Description
  ORDER BY
    Revenue DESC
  LIMIT 5
''', conn)

product_revenue

Description    Revenue
0                      DOTCOM POSTAGE  206245.48
1            REGENCY CAKESTAND 3 TIER  164762.19
2  WHITE HANGING HEART T-LIGHT HOLDER   99668.47
3                       PARTY BUNTING   98302.98
4             JUMBO BAG RED RETROSPOT   92356.03

In [19]:
for index, rows in product_revenue.iterrows():
  tuple_product = (rows['Description'], rows['Revenue'])
  sql_top_product_revenues.append(tuple_product)

sql_top_product_revenues

[('DOTCOM POSTAGE', 206245.48000000019),
 ('REGENCY CAKESTAND 3 TIER', 164762.19),
 ('WHITE HANGING HEART T-LIGHT HOLDER', 99668.46999999849),
 ('PARTY BUNTING', 98302.98000000048),
 ('JUMBO BAG RED RETROSPOT', 92356.03000000204)]

#### 3. Identify the most profitable countries (by total sales), limited to top 5

In [20]:
top_countries = pd.read_sql_query('''
  SELECT
    Country,
    SUM(Quantity * UnitPrice) as Revenue
  FROM
    ecommerce
  GROUP BY
    Country
  ORDER BY
    Revenue DESC
  LIMIT 5
''', conn)

top_countries

Country       Revenue
0  United Kingdom  8.187806e+06
1     Netherlands  2.846615e+05
2            EIRE  2.632768e+05
3         Germany  2.216982e+05
4          France  1.974039e+05

In [21]:
for index, rows in top_countries.iterrows():
  tuple_country = (rows['Country'], rows['Revenue'])
  sql_top_countries.append(tuple_country)

sql_top_countries

[('United Kingdom', 8187806.364001113),
 ('Netherlands', 284661.54000000015),
 ('EIRE', 263276.81999999826),
 ('Germany', 221698.20999999862),
 ('France', 197403.90000000002)]

#### 4. Find the top-selling products (by total sales) for each country, limited to top 5

In [22]:
# Example: sql_top_selling_products_by_country_by_sales = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 200000.00), ('COUNTRY 1 PRODUCT 2', 100000.00), ('COUNTRY 1 PRODUCT 3', 95348.00), ('COUNTRY 1 PRODUCT 4', 90000.43), ('COUNTRY 1 PRODUCT 5', 80000.00)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000.00), ('COUNTRY 2 PRODUCT 2', 7345.00), ('COUNTRY 2 PRODUCT 3', 6934.23), ('COUNTRY 2 PRODUCT 4', 6800.23), ('COUNTRY 2 PRODUCT 5', 4000.00)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 7442.84), ('COUNTRY 5 PRODUCT 2', 5175.00), ('COUNTRY 5 PRODUCT 3', 3098.0), ('COUNTRY 5 PRODUCT 4', 3014.34), ('COUNTRY 5 PRODUCT 5', 2857.83)],

In [23]:
top_product_countries = pd.read_sql_query('''
SELECT
    rp.Country,
    rp.Description,
    rp.Revenue,
    rp.ProductRank
FROM (
    SELECT
        rp_sub.Country,
        rp_sub.Description,
        rp_sub.Revenue,
        rp_sub.ProductRank,
        rc.CountryRank
    FROM (
        SELECT
            Country,
            Description,
            SUM(Quantity * UnitPrice) AS Revenue,
            RANK() OVER (PARTITION BY Country ORDER BY SUM(Quantity * UnitPrice) DESC) AS ProductRank
        FROM
            ecommerce
        GROUP BY
            Country,
            Description
    ) rp_sub
    JOIN (
        SELECT
            Country AS CountryAlias,
            RANK() OVER (ORDER BY SUM(Quantity * UnitPrice) DESC) AS CountryRank
        FROM
            ecommerce
        GROUP BY
            Country
    ) rc ON rp_sub.Country = rc.CountryAlias
    WHERE
        rc.CountryRank <= 5
) rp
WHERE
    rp.ProductRank <= 5
ORDER BY
    rp.CountryRank,
    rp.ProductRank;

''', conn)

top_product_countries

Country                          Description    Revenue  \
0   United Kingdom                       DOTCOM POSTAGE  206245.48   
1   United Kingdom             REGENCY CAKESTAND 3 TIER  134405.94   
2   United Kingdom   WHITE HANGING HEART T-LIGHT HOLDER   93953.07   
3   United Kingdom                        PARTY BUNTING   92501.73   
4   United Kingdom              JUMBO BAG RED RETROSPOT   84516.44   
5      Netherlands                   RABBIT NIGHT LIGHT    9568.48   
6      Netherlands  ROUND SNACK BOXES SET OF4 WOODLAND     7991.40   
7      Netherlands                  SPACEBOY LUNCH BOX     7485.60   
8      Netherlands                 DOLLY GIRL LUNCH BOX    6828.60   
9      Netherlands   ROUND SNACK BOXES SET OF 4 FRUITS     4039.20   
10            EIRE             REGENCY CAKESTAND 3 TIER    7442.85   
11            EIRE                             CARRIAGE    5175.00   
12            EIRE             JAM MAKING SET WITH JARS    3089.00   
13            EIRE        3 TIER CAKE TIN RED AND CREAM    3041.55   
14            EIRE   WHITE HANGING HEART T-LIGHT HOLDER    2857.80   
15         Germany                              POSTAGE   20821.00   
16         Germany             REGENCY CAKESTAND 3 TIER    8257.35   
17         Germany  ROUND SNACK BOXES SET OF4 WOODLAND     3554.70   
18         Germany   ROUND SNACK BOXES SET OF 4 FRUITS     1949.95   
19         Germany                  SPACEBOY LUNCH BOX     1629.45   
20          France                              POSTAGE   15065.00   
21          France                   RABBIT NIGHT LIGHT    7275.12   
22          France             REGENCY CAKESTAND 3 TIER    2581.80   
23          France        RED TOADSTOOL LED NIGHT LIGHT    2168.10   
24          France     PLASTERS IN TIN WOODLAND ANIMALS    1868.40   

    ProductRank  
0             1  
1             2  
2             3  
3             4  
4             5  
5             1  
6             2  
7             3  
8             4  
9             5  
10            1  
11            2  
12            3  
13            4  
14            5  
15            1  
16            2  
17            3  
18            4  
19            5  
20            1  
21            2  
22            3  
23            4  
24            5

In [24]:
selected_country = ''

for index, rows in top_product_countries.iterrows():
  if rows["Country"] != selected_country:
    sql_top_selling_products_by_country_by_sales[rows["Country"]] = []
    selected_country = rows["Country"]
  tuple_country = (rows['Description'], rows['Revenue'])
  sql_top_selling_products_by_country_by_sales[rows["Country"]].append(tuple_country)

In [25]:
sql_top_selling_products_by_country_by_sales

{'United Kingdom': [('DOTCOM POSTAGE', 206245.48000000019),
  ('REGENCY CAKESTAND 3 TIER', 134405.93999999992),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 93953.0699999987),
  ('PARTY BUNTING', 92501.73000000033),
  ('JUMBO BAG RED RETROSPOT', 84516.44000000166)],
 'Netherlands': [('RABBIT NIGHT LIGHT', 9568.48),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 7991.400000000001),
  ('SPACEBOY LUNCH BOX ', 7485.5999999999985),
  ('DOLLY GIRL LUNCH BOX', 6828.599999999999),
  ('ROUND SNACK BOXES SET OF 4 FRUITS ', 4039.2000000000007)],
 'EIRE': [('REGENCY CAKESTAND 3 TIER', 7442.849999999997),
  ('CARRIAGE', 5175.0),
  ('JAM MAKING SET WITH JARS', 3089.0),
  ('3 TIER CAKE TIN RED AND CREAM', 3041.5499999999997),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 2857.8)],
 'Germany': [('POSTAGE', 20821.0),
  ('REGENCY CAKESTAND 3 TIER', 8257.349999999997),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3554.6999999999985),
  ('ROUND SNACK BOXES SET OF 4 FRUITS ', 1949.9500000000016),
  ('SPACEBOY LUNCH BOX ', 

#### 5. Find the top-selling products (by quantity) for each country, limited to top 5

In [26]:
# Example: sql_top_selling_products_by_country_by_quantity = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 22343), ('COUNTRY 1 PRODUCT 2', 12345), ('COUNTRY 1 PRODUCT 3', 9534), ('COUNTRY 1 PRODUCT 4', 9000), ('COUNTRY 1 PRODUCT 5', 8000)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000), ('COUNTRY 2 PRODUCT 2', 7345), ('COUNTRY 2 PRODUCT 3', 6934), ('COUNTRY 2 PRODUCT 4', 6800), ('COUNTRY 2 PRODUCT 5', 4000)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 48345), ('COUNTRY 5 PRODUCT 2', 43965), ('COUNTRY 5 PRODUCT 3', 43556), ('COUNTRY 5 PRODUCT 4', 34567), ('COUNTRY 5 PRODUCT 5', 33436)],




In [27]:
top_product_countries = pd.read_sql_query('''
SELECT
    rp.Country,
    rp.Description,
    rp.Quantity,
    rp.ProductRank
FROM (
    SELECT
        rp_sub.Country,
        rp_sub.Description,
        rp_sub.Quantity,
        rp_sub.ProductRank,
        rc.CountryRank
    FROM (
        SELECT
            Country,
            Description,
            SUM(Quantity) as Quantity,
            RANK() OVER (PARTITION BY Country ORDER BY SUM(Quantity) DESC) AS ProductRank
        FROM
            ecommerce
        GROUP BY
            Country,
            Description
    ) rp_sub
    JOIN (
        SELECT
            Country AS CountryAlias,
            RANK() OVER (ORDER BY SUM(Quantity) DESC) AS CountryRank
        FROM
            ecommerce
        GROUP BY
            Country
    ) rc ON rp_sub.Country = rc.CountryAlias
    WHERE
        rc.CountryRank <= 5
) rp
WHERE
    rp.ProductRank <= 5
ORDER BY
    rp.CountryRank,
    rp.ProductRank;

''', conn)

top_product_countries

Country                          Description  Quantity  ProductRank
0   United Kingdom    WORLD WAR 2 GLIDERS ASSTD DESIGNS     48326            1
1   United Kingdom              JUMBO BAG RED RETROSPOT     43167            2
2   United Kingdom                       POPCORN HOLDER     34365            3
3   United Kingdom        ASSORTED COLOUR BIRD ORNAMENT     33679            4
4   United Kingdom   WHITE HANGING HEART T-LIGHT HOLDER     33193            5
5      Netherlands                   RABBIT NIGHT LIGHT      4801            1
6      Netherlands                  SPACEBOY LUNCH BOX       4528            2
7      Netherlands                 DOLLY GIRL LUNCH BOX      4132            3
8      Netherlands      PACK OF 72 RETROSPOT CAKE CASES      4128            4
9      Netherlands  ROUND SNACK BOXES SET OF4 WOODLAND       3132            5
10            EIRE      PACK OF 72 RETROSPOT CAKE CASES      1728            1
11            EIRE          60 TEATIME FAIRY CAKE CASES      1536            2
12            EIRE                   VINTAGE SNAP CARDS      1492            3
13            EIRE                ASSORTED INCENSE PACK      1440            4
14            EIRE   PACK OF 60 PINK PAISLEY CAKE CASES      1296            5
15         Germany  ROUND SNACK BOXES SET OF4 WOODLAND       1218            1
16         Germany            ASSORTED COLOURS SILK FAN      1164            2
17         Germany                              POSTAGE      1104            3
18         Germany               WOODLAND CHARLOTTE BAG      1019            4
19         Germany      PACK OF 72 RETROSPOT CAKE CASES      1002            5
20          France                   RABBIT NIGHT LIGHT      4023            1
21          France              MINI PAINT SET VINTAGE       2196            2
22          France        RED TOADSTOOL LED NIGHT LIGHT      1314            3
23          France          SET/6 RED SPOTTY PAPER CUPS      1272            4
24          France        ASSORTED COLOUR BIRD ORNAMENT      1204            5

In [28]:
selected_country = ''

for index, rows in top_product_countries.iterrows():
  if rows["Country"] != selected_country:
    sql_top_selling_products_by_country_by_quantity[rows["Country"]] = []
    selected_country = rows["Country"]
  tuple_country = (rows['Country'], rows['Description'], rows['Quantity'])
  sql_top_selling_products_by_country_by_quantity[rows["Country"]].append(tuple_country)

In [29]:
sql_top_selling_products_by_country_by_quantity

{'United Kingdom': [('United Kingdom',
   'WORLD WAR 2 GLIDERS ASSTD DESIGNS',
   48326),
  ('United Kingdom', 'JUMBO BAG RED RETROSPOT', 43167),
  ('United Kingdom', 'POPCORN HOLDER', 34365),
  ('United Kingdom', 'ASSORTED COLOUR BIRD ORNAMENT', 33679),
  ('United Kingdom', 'WHITE HANGING HEART T-LIGHT HOLDER', 33193)],
 'Netherlands': [('Netherlands', 'RABBIT NIGHT LIGHT', 4801),
  ('Netherlands', 'SPACEBOY LUNCH BOX ', 4528),
  ('Netherlands', 'DOLLY GIRL LUNCH BOX', 4132),
  ('Netherlands', 'PACK OF 72 RETROSPOT CAKE CASES', 4128),
  ('Netherlands', 'ROUND SNACK BOXES SET OF4 WOODLAND ', 3132)],
 'EIRE': [('EIRE', 'PACK OF 72 RETROSPOT CAKE CASES', 1728),
  ('EIRE', '60 TEATIME FAIRY CAKE CASES', 1536),
  ('EIRE', 'VINTAGE SNAP CARDS', 1492),
  ('EIRE', 'ASSORTED INCENSE PACK', 1440),
  ('EIRE', 'PACK OF 60 PINK PAISLEY CAKE CASES', 1296)],
 'Germany': [('Germany', 'ROUND SNACK BOXES SET OF4 WOODLAND ', 1218),
  ('Germany', 'ASSORTED COLOURS SILK FAN', 1164),
  ('Germany', 'POSTAGE

In [30]:
conn.close()

In [31]:
question_id = "01_sql_columns"
submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
question_id = "02_sql_num_rows"
submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
question_id = "03_sql_num_products"
submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
question_id = "04_sql_num_customers"
submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
question_id = "05_sql_top_product_revenues"
submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
question_id = "06_sql_top_countries"
submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
question_id = "07_sql_top_selling_products_by_country_by_sales"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
question_id = "08_sql_top_selling_products_by_country_by_quantity"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)

'Assignment successfully submitted'

## II. Indexing and Querying with Elasticsearch

You should have a better understanding of how to use SQL, now let's leverage the power of Elasticsearch to explore and analyze the data. We'll use the same dataset.


### II.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [32]:
!pip install elasticsearch==7.9.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 213.5/213.5 kB 3.9 MB/s eta 0:00:00


In [33]:
from elasticsearch import Elasticsearch
from elasticsearch.helpers import bulk
import pandas as pd
import time
import json

import warnings
warnings.filterwarnings("ignore")

In [34]:
%%bash

wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.2-linux-x86_64.tar.gz
wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.2-linux-x86_64.tar.gz.sha512
tar -xzf elasticsearch-oss-7.9.2-linux-x86_64.tar.gz
sudo chown -R daemon:daemon elasticsearch-7.9.2/
shasum -a 512 -c elasticsearch-oss-7.9.2-linux-x86_64.tar.gz.sha512

elasticsearch-oss-7.9.2-linux-x86_64.tar.gz: OK


In [35]:
%%bash --bg

sudo -H -u daemon elasticsearch-7.9.2/bin/elasticsearch

In [36]:
time.sleep(20)

In [37]:
%%bash

ps -ef | grep elasticsearch

root        1535    1533  0 16:30 ?        00:00:00 sudo -H -u daemon elasticsearch-7.9.2/bin/elasti
daemon      1536    1535 78 16:30 ?        00:00:15 /content/elasticsearch-7.9.2/jdk/bin/java -Xshar
root        1838    1836  0 16:30 ?        00:00:00 grep elasticsearch


In [39]:
%%bash

curl -sX GET "localhost:9200/"

{
  "name" : "e6a6e5210f75",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "fKotaOehSem27yVBb2vPFA",
  "version" : {
    "number" : "7.9.2",
    "build_flavor" : "oss",
    "build_type" : "tar",
    "build_hash" : "d34da0ea4a966c4e49417f2da2f244e3e97b4e6e",
    "build_date" : "2020-09-23T00:45:33.626720Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}


### II.2. Set up Elasticsearch Index

The first step you need to do which is **preprocessing, let's keep it simple and simply drop all row that have missing values.**

Next, you need to set up an Elasticsearch index. An Elasticsearch index is like a database in a traditional relational database. An index lets you store, search, and analyze big volumes of data quickly and in near real time.

You'll also need to setup the mappings for the data types, let's do that here.

In [40]:
df_elastic = df.copy()

In [41]:
# Convert 'InvoiceDate' column to datetime format
df_elastic['InvoiceDate'] = pd.to_datetime(df_elastic['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Add leading zero to month if it's a single digit
df_elastic['InvoiceDate'] = df_elastic['InvoiceDate'].dt.strftime('%m/%d/%Y %H:%M')

In [42]:
df_elastic.head(3)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   
2    536365    84406B      CREAM CUPID HEARTS COAT HANGER         8   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom

In [43]:
df_elastic.isna().sum().sum()

136534

In [44]:
# Drop rows with missing values
df_elastic.dropna(inplace=True)

In [45]:
df_elastic.isna().sum().sum()

0

### II.3. Verifying the index

Before we go to the next section, let's check our index first. You need to change the my_index_name with your chosen index name, afterwards you can just run the code block. The result should be:

- Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
- Number of rows: 406829

In [46]:
# Connect to the Elasticsearch server
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

In [47]:
# Define index name
index_name = 'ecommerce'

In [48]:
# Define index settings and mappings
index_settings = {
    'mappings': {
        'properties': {
            'InvoiceNo': {'type': 'keyword'},
            'StockCode': {'type': 'keyword'},
            'Description': {'type': 'keyword'},
            'Quantity': {'type': 'integer'},
            'InvoiceDate': {'type': 'date', 'format': 'MM/dd/yyyy HH:mm'},
            'UnitPrice': {'type': 'float'},
            'CustomerID': {'type': 'float'},
            'Country': {'type': 'keyword'}
        }
    }
}

In [49]:
# # Delete the index
# response_delete = es.indices.delete(index=index_name, ignore=[400, 404])

# # Check if the index was deleted successfully
# if 'acknowledged' in response_delete and response_delete['acknowledged']:
#     print(f"Index '{index_name}' deleted successfully.")
# else:
#     print(f"Failed to delete index '{index_name}'.")

In [50]:
# Create the index with the specified settings and mappings
response_create = es.indices.create(index=index_name, body=index_settings, ignore=400)
print(json.dumps(response_create, indent=4))

{
    "acknowledged": true,
    "shards_acknowledged": true,
    "index": "ecommerce"
}


In [51]:
# Convert DataFrame to a list of dictionaries
docs = df_elastic.to_dict(orient='records')

In [52]:
# Function to generate Elasticsearch actions
def generate_actions(docs):
    for doc in docs:
        yield {
            "_index": index_name,
            "_source": doc
        }

In [53]:
success, _ = bulk(es, generate_actions(docs))
print(f"Inserted {success} documents into the index '{index_name}'.")

Inserted 406829 documents into the index 'ecommerce'.


#### Get Result

In [54]:
# Get index mapping (equivalent to getting column names in SQL)
mapping = es.indices.get_mapping(index=index_name)
es_columns = list(mapping[index_name]['mappings']['properties'].keys()) #the variable es_columns is used for submission, do not change the variable name
print(f"Columns: {es_columns}")

# Count documents in the index (equivalent to counting rows in SQL)
es_num_rows = es.count(index=index_name)['count'] #the variable es_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {es_num_rows}")


Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
Number of rows: 406829


### II.4. Running Elasticsearch queries

Once your e-commerce data is cleaned and ready, let's analyse the data to gain insights that will drive the expansion strategy, this time using Elasticsearch and we'll focus on one specific country, which is Germany.

1. **Find all transactions for a specific country**

   Your company operates globally, but you want to understand better how different countries contribute to your sales. By finding all transactions that happened in a specific country, you can get a clearer picture of your company's reach and performance in that location. In our case, we'll focus on the country Germany.


2. **Find all unique products in Germany**

   Your first task is to understand the breadth of your company's product portfolio in Germany. By determining the total number of unique products you sell, you can assess the scale and diversity of your operations.


3. **Find the top 5 most purchased products in Germany**

   Next up, you'll identify the best-selling products in your portfolio in Germany. Knowing which items are the most purchased can help guide decisions about product focus and marketing efforts. We'll limit the result to 5 most purchased products.


By executing these tasks, you will enable your team to provide data-driven insights which will directly contribute to the strategic decisions about your business expansion. As part of the data engineering team, you're not only maintaining the technical infrastructure - you're shaping the future of the company.

There are two more tasks below to check your understanding on how to handle date type data as well as searching using partial keyword:

4. **Find a product's transactions over time**

   Trends in purchases over time for specific products can provide insights into product popularity and seasonality. You decide to track a product's transactions over a period to understand its sales pattern. In this case we'll use the product "I LOVE LONDON MINI RUCKSACK"

5. **Search for a product using a part of its description**

   Finally, you want to investigate how robust and reliable your product search feature is. You decide to test it by searching for a product using a part of its description. In this case, we'll use the keyword "PIN". Return only unique product name. And the product names should be stored in a list in alphabetical order.


In [55]:
# Write your Elasticsearch Query here

#use the following variable name to store the result from your Elasticsearch Query
es_total_unique_products = 0
es_transactions_in_germany = 0
es_top_products = {} #dictionary where key is product name, value is the quantity sold, see example below
es_transactions_over_time = {} #dictionary where key is invoice date, value is the quantity sold, see example below
es_unique_product_search = [] #list containing the products in alphabetical order

#### 1. Find all transactions for a specific country (Germany)

Example: es_transactions_in_germany = 7584

In [56]:
# Define the query
query = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    }
}

# Execute the query
response = es.search(index=index_name, body=query)

# Get the count of matching transactions
es_transactions_in_germany = response["hits"]["total"]["value"]
print("Total transactions in Germany:", es_transactions_in_germany)

Total transactions in Germany: 9495


#### 2. Find all unique products in Germany

Hints: Use set() to remove duplicate values

Example: es_total_unique_products = 45345

In [57]:
# Define the query to retrieve all documents with country "Germany"
query = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    },
    "size": 10000  # Set size to a large number to ensure retrieval of all documents
}

# Execute the query
response = es.search(index=index_name, body=query)

# Extract unique products from the response
unique_products = set()
for hit in response["hits"]["hits"]:
    unique_products.add(hit["_source"]["Description"])

# Get the total count of unique products
es_total_unique_products = len(unique_products)
print("Total unique products in Germany:", es_total_unique_products)

Total unique products in Germany: 1703


#### 3. Find the top 5 most purchased products in Germany

Example: es_top_products = {'PRODUCT 1': 498, 'PRODUCT 2': 452, 'PRODUCT 3': 342, 'PRODUCT 4': 231, 'PRODUCT 5': 123}

In [58]:
# Define the query to retrieve top 5 most purchased products in Germany
query = {
    "size": 0,
    "query": {
        "match": {
            "Country": "Germany"  # Filter for transactions in Germany
        }
    },
    "aggs": {
        "top_products": {
            "terms": {
                "field": "Description",
                "size": 5
            },
            "aggs": {
                "count": {
                    "sum": {
                        "field": "Quantity"
                    }
                }
            }
        }
    }
}

In [59]:
response = es.search(index=index_name, body=query)

top_products = {}
for bucket in response['aggregations']['top_products']['buckets']:
    product_name = bucket['key']
    count = bucket['count']['value']
    top_products[product_name] = count

es_top_products = top_products
print(es_top_products)

{'POSTAGE': 1104.0, 'ROUND SNACK BOXES SET OF4 WOODLAND ': 1218.0, 'REGENCY CAKESTAND 3 TIER': 737.0, 'ROUND SNACK BOXES SET OF 4 FRUITS ': 661.0, 'PLASTERS IN TIN WOODLAND ANIMALS': 842.0}


#### 4. Find a product's transactions over time

Hints: Date format is M/d/yyyy H:m

Product = I LOVE LONDON MINI RUCKSACK

Example: es_transactions_over_time = {'12/12/2010 11:11': 1, 12/12/2010 11:12': 1}

In [60]:
from datetime import datetime

product_name = " I LOVE LONDON MINI RUCKSACK"

query = {
    "size": 10000,  # Set size to a large number to ensure retrieval of all relevant documents
    "query": {
        "match": {
            "Description": product_name
        }
    }
}

In [61]:
response = es.search(index=index_name, body=query)

transactions_over_time = {}
for hit in response["hits"]["hits"]:
    timestamp = datetime.strptime(hit["_source"]["InvoiceDate"], "%m/%d/%Y %H:%M")
    timestamp_str = timestamp.strftime("%m/%d/%Y %H:%M")  # Format output as "M/d/yyyy H:m"
    transactions_over_time[timestamp_str] = transactions_over_time.get(timestamp_str, 0) + 1

es_transactions_over_time = transactions_over_time
print(es_transactions_over_time)

{'10/06/2011 09:11': 1}


#### 5. Search for a product using a part of its description

Hints: Return only unique Product names, see number 1

Example: es_unique_product_search = ['PRODUCT 1', 'PRODUCT 2']


In [62]:
search_term = "PIN "  # Lowercase search term

query = {
    "size": 0,  # We only need aggregations, no need to retrieve documents
    "query": {
        "wildcard": {
            "Description": f"*{search_term.upper()}*"  # Using uppercase to match any case of "PIN"
        }
    },
    "aggs": {
        "unique_descriptions": {
            "terms": {
                "field": "Description",  # Perform aggregation on the "Description" field
                "size": 10,  # Adjust the size according to your needs
                "order": {
                    "_key": "asc"  # Sort the unique descriptions alphabetically
                }
            }
        }
    }
}

In [63]:
response = es.search(index=index_name, body=query)

# Extract unique product names from the response
es_unique_product_search = [bucket['key'] for bucket in response['aggregations']['unique_descriptions']['buckets']]
es_unique_product_search

['PANTRY 3 HOOK ROLLING PIN HANGER',
 'PIN CUSHION BABUSHKA BLUE',
 'PIN CUSHION BABUSHKA PINK',
 'PIN CUSHION BABUSHKA RED']

## Submission

Once you are satisfied with the performance of your model, then you run the code block below to submit your project.


In [64]:
# Submit Method

# Do not change the code below
question_id = "01_sql_columns"
submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
question_id = "02_sql_num_rows"
submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
question_id = "03_sql_num_products"
submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
question_id = "04_sql_num_customers"
submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
question_id = "05_sql_top_product_revenues"
submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
question_id = "06_sql_top_countries"
submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
question_id = "07_sql_top_selling_products_by_country_by_sales"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
question_id = "08_sql_top_selling_products_by_country_by_quantity"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)

question_id = "09_es_columns"
submit(student_id, name, assignment_id, str(es_columns), question_id, drive_link)
question_id = "10_es_num_rows"
submit(student_id, name, assignment_id, str(es_num_rows), question_id, drive_link)
question_id = "11_es_total_unique_products"
submit(student_id, name, assignment_id, str(es_total_unique_products), question_id, drive_link)
question_id = "12_es_transactions_in_germany"
submit(student_id, name, assignment_id, str(es_transactions_in_germany), question_id, drive_link)
question_id = "13_es_top_products"
submit(student_id, name, assignment_id, str(es_top_products), question_id, drive_link)
question_id = "14_es_transactions_over_time"
submit(student_id, name, assignment_id, str(es_transactions_over_time), question_id, drive_link)
question_id = "15_es_unique_product_search"
submit(student_id, name, assignment_id, str(es_unique_product_search), question_id, drive_link)

'Assignment successfully submitted'

## FIN